### PR TITLE
Prune options properly instead of retaining them all

### PR DESCRIPTION
**Pull request by oldergod**
Part of #1243 Let's go !

### CODE REVIEW

File: wire-schema/src/main/java/com/squareup/wire/schema/Options.kt [hunk 1]
```diff
@@ -36,17 +37,24 @@ import kotlin.collections.set
...
-  val elements: ImmutableList<OptionElement> = ImmutableList.copyOf(elements)
-
   // Null until this options is linked.
-  var map: ImmutableMap<ProtoMember, Any?>? = null
-    private set
+  private var entries: List<LinkedOptionEntry>? = null
+
+  val elements: List<OptionElement>
+    get() {
+      return entries?.map { it.optionElement } ?: optionElements
```
Comment (swankjesse) [inline wire-schema/src/main/java/com/squareup/wire/schema/Options.kt]: neat!

File: wire-schema/src/main/java/com/squareup/wire/schema/Options.kt [hunk 2]
```diff
@@ -55,26 +63,30 @@ class Options(
...
-    for (option in elements) {
-      val canonicalOption = canonicalizeOption(linker, optionType, option) ?: continue
-      map = union(linker, map, canonicalOption)
+    var entries: List<LinkedOptionEntry> = emptyList()
+
+    for (option in optionElements) {
+      val canonicalOption: List<LinkedOptionEntry> = canonicalizeOption(linker, optionType, option)
+              ?: continue
+
+      entries = union(linker, entries, canonicalOption)
```
Comment (swankjesse) [inline wire-schema/src/main/java/com/squareup/wire/schema/Options.kt]: For what it’s worth, this is `O(N*N)` which is bad if anyone has lots of options

File: wire-schema/src/main/java/com/squareup/wire/schema/Options.kt [hunk 3]
```diff
@@ -198,6 +214,33 @@ class Options(
...
+    for (entry in bMap) {
+      val bValue = entry.value
+      val aValue = result[entry.key]
+      result[entry.key] = when {
+        aValue != null -> union(linker, aValue, bValue)
+        else -> bValue
+      }
+    }
+
+    return a.map { it.optionElement to it.protoMember }
```
Comment (swankjesse) [inline wire-schema/src/main/java/com/squareup/wire/schema/Options.kt]: this reads really awkwardly, especially as we have a.map and aMap

File: wire-schema/src/main/java/com/squareup/wire/schema/Options.kt [hunk 4]
```diff
@@ -198,6 +214,33 @@ class Options(
     }
   }
 
+  private fun union(
+    linker: Linker, a: List<LinkedOptionEntry>, b: List<LinkedOptionEntry>
+  ): List<LinkedOptionEntry> {
+    val aMap: Map<ProtoMember, Any?> = a.toMap()
+    val bMap: Map<ProtoMember, Any> = b.toMap() as Map<ProtoMember, Any>
+
+    val result: MutableMap<ProtoMember, Any?> = LinkedHashMap(aMap)
```
Comment (swankjesse) [inline wire-schema/src/main/java/com/squareup/wire/schema/Options.kt]: should this be named result?

File: wire-schema/src/main/java/com/squareup/wire/schema/Options.kt [hunk 5]
```diff
@@ -256,12 +297,20 @@ class Options(
   }
 
   fun retainAll(schema: Schema, markSet: MarkSet): Options {
-    if (map == null || map!!.isEmpty()) return this // Nothing to prune.
+    if (entries.isNullOrEmpty()) return this // Nothing to prune.
```
Comment (swankjesse) [inline wire-schema/src/main/java/com/squareup/wire/schema/Options.kt]: is this invoked in both null and empty cases?

File: wire-schema/src/main/java/com/squareup/wire/schema/Options.kt [hunk 6]
```diff
@@ -353,4 +402,8 @@ class Options(
       return null
     }
   }
+
+  private fun List<LinkedOptionEntry>.toMap(): Map<ProtoMember, Any?> {
```
Comment (swankjesse) [inline wire-schema/src/main/java/com/squareup/wire/schema/Options.kt]: do you mean associateBy ?

File: wire-schema/src/main/java/com/squareup/wire/schema/Schema.java [hunk 7]
```diff
@@ -125,12 +132,19 @@ static Schema fromFiles(Iterable<ProtoFile> sourceProtoFiles, Loader pathFilesLo
...
 
-  private static ImmutableMap<String, Type> buildTypesIndex(Iterable<ProtoFile> protoFiles) {
+  private static ImmutableMap<String, Type> buildTypesIndex(Iterable<ProtoFile> protoFiles,
+      Map<ProtoType, ProtoFile> protoFilesIndex) {
     Map<String, Type> result = new LinkedHashMap<>();
     for (ProtoFile protoFile : protoFiles) {
       for (Type type : protoFile.getTypes()) {
         index(result, type);
       }
+      for (Type type : result.values()) {
```
Comment (swankjesse) [inline wire-schema/src/main/java/com/squareup/wire/schema/Schema.java]: this loop should be out outside the enclosing for loop

Comment (swankjesse) [reply wire-schema/src/main/java/com/squareup/wire/schema/Schema.java]: I don’t think this is the best way to index these

Comment (swankjesse) [reply wire-schema/src/main/java/com/squareup/wire/schema/Schema.java]: ... instead, make the `index()` method take the `protoFilesIndex` ?

File: wire-schema/src/main/java/com/squareup/wire/schema/internal/parser/LinkedOptionEntry.kt [hunk 8]
```diff
@@ -0,0 +1,9 @@
+package com.squareup.wire.schema.internal.parser
+
+import com.squareup.wire.schema.ProtoMember
+
+data class LinkedOptionEntry(
```
Comment (swankjesse) [inline wire-schema/src/main/java/com/squareup/wire/schema/internal/parser/LinkedOptionEntry.kt]: Is this in the wrong package?

**Comment by swankjesse** [comment]
I’m a little confused by the flipping between Lists and Maps.

File: wire-schema/src/main/java/com/squareup/wire/schema/Schema.java (hunk 7 above)
Comment (oldergod) [reply wire-schema/src/main/java/com/squareup/wire/schema/Schema.java]: How could I get my hands on the right protoFile if I'm outside the enclosing loop?

File: wire-schema/src/main/java/com/squareup/wire/schema/Options.kt (hunk 5 above)
Comment (oldergod) [reply wire-schema/src/main/java/com/squareup/wire/schema/Options.kt]: yes

File: wire-schema/src/main/java/com/squareup/wire/schema/Options.kt (hunk 3 above)
Comment (oldergod) [reply wire-schema/src/main/java/com/squareup/wire/schema/Options.kt]: I agree

File: wire-schema/src/main/java/com/squareup/wire/schema/Options.kt (hunk 2 above)
Comment (oldergod) [reply wire-schema/src/main/java/com/squareup/wire/schema/Options.kt]: We can look to optimise later. It's already N*N I believe.

File: wire-schema/src/main/java/com/squareup/wire/schema/Schema.java (hunk 7 above)
Comment (oldergod) [reply wire-schema/src/main/java/com/squareup/wire/schema/Schema.java]: I can pass both the protoFile and the `protoFilesIndex` to `index()`

**Comment by oldergod** [comment]
> I’m a little confused by the flipping between Lists and Maps.

Yes this sucks. I cheated by creating an early barrier between the change I needed and the refactoring that is much required.

File: proto-pruner/src/test/java/com/squareup/wire/prune/ProtoPrunerTest.kt [hunk 9]
```diff
@@ -97,7 +96,8 @@ class ProtoPrunerTest {
     invokeProtoPruner(identifierSet)
 
     val outputs = arrayOf(
-        "poem.proto"
+        "poem.proto",
+        "options.proto"
```
Comment (oldergod) [inline proto-pruner/src/test/java/com/squareup/wire/prune/ProtoPrunerTest.kt]: this is a bug I wanna fix later.

**Comment by swankjesse** [comment]
Glad this works. Let's land and iterate!

**Pull request merged**